File: C++/2458. Height of Binary Tree After Subtree Removal Queries.cpp
class Solution {
public:
    map<int, int> lh, rh, ht; 
    int height(TreeNode* root) {
        if (!root) return 0;
        int lr = height(root->left);  
        int rr = height(root->right); 
        lh[root->val] = lr; 
        rh[root->val] = rr;  
        return max(lr, rr) + 1;  
    }
    void solve(TreeNode* root, int maxHeightAbove, int depth) {
        if (!root) return;
        ht[root->val] = maxHeightAbove; 
        solve(root->left, max(maxHeightAbove, depth + rh[root->val]), depth + 1);
        solve(root->right, max(maxHeightAbove, depth + lh[root->val]), depth + 1);
    }
    vector<int> treeQueries(TreeNode* root, vector<int>& queries) {
        height(root); 
        solve(root->left, rh[root->val], 1); 
        solve(root->right, lh[root->val], 1); 
        vector<int> ans;
        for (int x : queries) {
            ans.push_back(ht[x]);  
        }
        return ans; 
    }
};

File: JAVA/2458. Height of Binary Tree After Subtree Removal Queries.java
class Solution {
    int[] height = new int[1000001];
    int[] level = new int[1000001];
    int[] high = new int[1000001];
    int[] secondHigh = new int[1000001];
    public int[] treeQueries(TreeNode root, int[] queries) {
        height(root,0);
        int[] ans = new int[queries.length];
        for(int i=0;i<queries.length;i++)
        {
            int node = queries[i];
            int l = level[node];
            int temp = l+(high[l]==height[node] ? secondHigh[l] : high[l])-1;
            ans[i] = temp;
        }
        return ans;
    }
    private int height(TreeNode root,int l)
    {
        if(root==null)
        return 0;

        level[root.val]=l;
        height[root.val] = Math.max(height(root.left,l+1),height(root.right,l+1))+1;

        if(high[l]<height[root.val])
        {
            secondHigh[l] = high[l];
            high[l]=height[root.val];
        }
        else if(secondHigh[l]<height[root.val])
        {
            secondHigh[l]=height[root.val];
        }
        return height[root.val];
    }
}
